### PR TITLE
bitpit: fixes for some misusage related to references

### PR DESCRIPTION
**Pull request by andrea-iob**
Fixes the following type of misusages of references:
 - small and trivially-copyable types passed by reference;
 - large types passed by value;
 - usage of copies in for-range loops.

### CODE REVIEW

File: src/containers/piercedStorage.hpp [hunk 1]
```diff
@@ -405,7 +405,7 @@ friend class PiercedStorageIterator;
     template<typename T = value_t, typename std::enable_if<PiercedStorage<T, id_t>::has_restore()>::type * = nullptr>
     void restoreField(std::istream &stream, T &object);
 
-    void dumpField(std::ostream &stream, const std::vector<bool>::const_reference &value) const;
+    void dumpField(std::ostream &stream, std::vector<bool>::const_reference value) const;
```
Comment (edoardolombardi) [inline src/containers/piercedStorage.hpp]: Why not just bool instead of vector::const_reference?

File: src/IO/DGF.cpp [hunk 2]
```diff
@@ -107,7 +107,7 @@ return; }
     \param[in] filename dgf file name
 */
 DGFObj::DGFObj(
-    std::string                          filename
+    const std::string                   &filename
```
Comment (edoardolombardi) [inline src/IO/DGF.cpp]: If I well understood the modifications (from copy value to reference) are not related to the title of the commit.

File: src/containers/piercedStorage.hpp (hunk 1 above)
Comment (andrea-iob) [reply src/containers/piercedStorage.hpp]: A std::vector<bool> may not be a vector of bool objects. To make it space-efficient, every implementation is free to chose any memory layout it likes to store the elements of that array. Hence a reference to an element of a std::vector<bool> may not be a bool.

File: src/IO/DGF.cpp (hunk 2 above)
Comment (andrea-iob) [reply src/IO/DGF.cpp]: In my view, "misusages" include also the cases where references should have been used but they were not (i.e., all the "*: pass large types by constant reference" commits).

Comment (edoardolombardi) [reply src/IO/DGF.cpp]: The title of the commit says "IO/DGF: pass small and trivially-copyable types by value". am I wrong ?

File: src/containers/piercedStorage.hpp (hunk 1 above)
Comment (edoardolombardi) [reply src/containers/piercedStorage.hpp]: Sorry, it's not really clear to me; my comment was related to std::vector<bool> element that I suppose you want to pass by value in this commit.

File: src/IO/DGF.cpp (hunk 2 above)
Comment (andrea-iob) [reply src/IO/DGF.cpp]: I though you meant the title of the pull request. I will fix the commit message.

File: src/containers/piercedStorage.hpp (hunk 1 above)
Comment (andrea-iob) [reply src/containers/piercedStorage.hpp]: The operator ```[]``` of a ```std::vector<bool>``` doesn't return a ```bool``` but a ```std::vector<bool>::reference```. 

Just to be sure, I will do some tests to check if, after the change, the above template specialization is still used when dumping a PiercedStorage of bools.

**Comment by andrea-iob** [comment]
I fixed the commit message for the IO/DGF changes and checked that the the dump of a PiercedStorage of bools works as expected.

**Comment by andrea-iob** [comment]
Can I merge this changes to the master branch?

**Pull request merged**